crates/fluvio/src/producer/dispatcher.rs

```diff
@@ -0,0 +1,236 @@
+use std::pin::Pin;
+use std::sync::Arc;
+use std::collections::HashMap;
+use futures_util::{StreamExt, Stream};
+use async_channel::Receiver;
+
+use tracing::{error, debug};
+use fluvio_types::SpuId;
+use fluvio_types::event::StickyEvent;
+use dataplane::ReplicaKey;
+use dataplane::batch::Batch;
+use dataplane::produce::{ProduceRequest, TopicProduceData, PartitionProduceData, ProduceResponse};
+
+use crate::FluvioError;
+use crate::spu::SpuPool;
+use crate::producer::{PendingRecord, DispatcherMessage};
+use crate::producer::partitioning::{Partitioner, SiphashRoundRobinPartitioner, PartitionerConfig};
+
+pub(crate) struct Dispatcher {
+    /// A pool of connections to the SPUs.
+    pool: Arc<SpuPool>,
+    /// The partitioning strategy to use
+    partitioner: Box<dyn Partitioner + Send + Sync>,
+    /// A buffer of records that have yet to be sent to the cluster.
+    buffer: Vec<PendingRecord>,
+    /// A stream of events that the dispatcher needs to react to.
+    ///
+    /// A `NewRecord` event means that the [`Producer`]'s caller has called
+    /// `send` to queue a new record, and we need to buffer it until we're ready
+    /// to send.
+    ///
+    /// A `Timeout` event means we need to flush the buffer if there is anything in it.
+    events: Option<Pin<Box<dyn Stream<Item = Event> + Send + Sync>>>,
+    /// A shutdown event used to determine when to quit.
+    ///
+    /// This will be triggered by the [`Producer`] that starts this dispatcher.
+    shutdown: StickyEvent,
+}
+
+enum Event {
+    ProducerRequest(DispatcherMessage),
+    Timeout,
+}
+
+impl Dispatcher {
+    pub(crate) fn new(pool: Arc<SpuPool>, incoming: Receiver<DispatcherMessage>) -> Self {
+        let partitioner = Box::new(SiphashRoundRobinPartitioner::new());
+        let shutdown = StickyEvent::new();
+
+        // Set up the event stream: Select between new records and timeouts
+        let timer =
```

Comment:
look at other dispatcher is done.  it's not good to pass stream in the future like this. it's better to create stream after spawning.  this willl avoid static lifetime